### PR TITLE
fix(apigateway): SAM CLI asset metadata missing from SpecRestApi

### LINKED ISSUE #14593: (assets, aws-lambda): Add asset metadata for nested stack and image-type Lambda Function
Add CloudFormation resource metadata for SAM to find local assets for the resources below:
1. local nested stack
2. image-type Lambda Function

### Use Case
In this [PR](https://github.com/aws/aws-cdk/pull/1433), resource metadata was added for zip-type Lambda Function for SAM to find local assets (for example, for running `sam local invoke`). As SAM CLI now supports local nested stacks and image-type Lambda Function, adding resource metadata for these two types of resources will enable SAM CLI to support CDK-synthed templates with these two types of resources.

### Proposed Solution

I believe the implementation will be very similar to the PR mentioned above.
For local nested stack, the value of `aws::asset::property` should be `"TemplateURL"`.
For image-type Lambda Function, it's a bit tricky - because the local path has to sit inside the `ImageUri` property under `Code`, I propose the value of `aws::asset::property` to be `"Code.ImageUri"`.

Example for local nested stack (in yaml):
```
nestedstackNestedStacknestedstackNestedStackResource71CDD241:
    Type: AWS::CloudFormation::Stack
    Properties:
      TemplateURL:
        Fn::Join:
          - ""
          - - https://s3.
            - Ref: AWS::Region
            - "."
            - Ref: AWS::URLSuffix
            - /
            - Ref: AssetParameters8cb8f69d88bba4ff29b6cfb1808b58a5d4b721166d137c367a744f8b6e5e65bfS3BucketFD698EA9
            - /
            - Fn::Select:
                - 0
                - Fn::Split:
                    - "||"
                    - Ref: AssetParameters8cb8f69d88bba4ff29b6cfb1808b58a5d4b721166d137c367a744f8b6e5e65bfS3VersionKey60183D77
            - Fn::Select:
                - 1
                - Fn::Split:
                    - "||"
                    - Ref: AssetParameters8cb8f69d88bba4ff29b6cfb1808b58a5d4b721166d137c367a744f8b6e5e65bfS3VersionKey60183D77
      Parameters:
        # ...
    UpdateReplacePolicy: Delete
    DeletionPolicy: Delete
    Metadata:
      aws:cdk:path: root-stack/nested-stack.NestedStack/nested-stack.NestedStackResource
      aws:asset:path: rootstacknestedstackACD02B51.nested.template.json # <- new
      aws:asset:property: TemplateURL # <- new
```

Example for image-type Lambda Function (in yaml):
```
containerfunction2D259B7A:
    Type: AWS::Lambda::Function
    Properties:
      Code:
        ImageUri:
          Fn::Join:
            - ""
            - - Ref: AWS::AccountId
              - .dkr.ecr.
              - Ref: AWS::Region
              - "."
              - Ref: AWS::URLSuffix
              - /aws-cdk/assets:58ff0b9a5ea87b6462d9a416b6066e346ae4b60f8f7ec6018972f50867181a9c
      Role:
        Fn::GetAtt:
          - containerfunctionServiceRole5EB80FD7
          - Arn
      ImageConfig:
        Command:
          - app.get
        EntryPoint:
          - /lambda-entrypoint.sh
      PackageType: Image
    DependsOn:
      - containerfunctionServiceRole5EB80FD7
    Metadata:
      aws:cdk:path: root-stack/container-function/Resource
      aws:asset:path: path/to/lambda_code # <- new
      aws:asset:property: Code.ImageUri # <- new
```

### Other

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by moelasmar**
Adds Assets metadata to RestApi resource in case if AssetApiDefinition is used. This Metadata will enable SAM
CLI to find local assets used by RestApi in the template.

It follows the same design in document [design/code-asset-metadata.md](https://github.com/aws/aws-cdk/pull/design/code-asset-metadata.md)

Fixes #14593

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by nija-at — DISMISSED** [review]
Can you explain what is going on with the changes to cdk-integ?

**Comment by moelasmar** [comment]
The integration between SAM and CDK was implemented before using [this design](https://github.com/aws/aws-cdk/blob/master/design/code-asset-metadata.md). The idea is SAM needs to know the local path of the assets attached to Lambda Function so SAM can implement the local testing and debugging functionalities. The solution was to let CDK add metadata properties for the Lambda Function resources to let SAM know the code path.

In This PR we want to extend this solution to SpecRestApi resource to add the assets metadata to it in case if it is using AssetApiDefinition.

**Comment by moelasmar** [comment]
in this PR .. I also did a change to enable the asset metadata generation in integration testing, and updated the Lambda function runtime from NodeJs10 to NodeJs14 to fix the integration testing failures.

File: packages/@aws-cdk/aws-apigateway/lib/api-definition.ts [hunk 1]
```diff
@@ -82,6 +83,15 @@ export abstract class ApiDefinition {
    * assume it's initialized. You may just use it as a construct scope.
    */
   public abstract bind(scope: Construct): ApiDefinitionConfig;
+
+  /**
+   * Called after the CFN RestApi resource has been created to allow the Api
+   * Definition to bind to it. Specifically it's required to allow assets to add
+   * metadata for tooling like SAM CLI to be able to find their origins.
+   */
+  public bindToResource(_resource: CfnResource) {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-apigateway/lib/api-definition.ts:92]: Can we make this: `bindAfterCreate(scope: Construct, restApi: RestApi)` ?

Call it like `apiDefinition.bindAfterCreate(this, this)`, and you can access the CfnResource inside like:

```ts
const child = Node.of(restApi).defaultChild as CfnRestApi;
child.addMetadata(...);
```
```

**Review by rix0rrr — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-apigateway/lib/api-definition.ts (hunk 1 above)
Comment (moelasmar) [reply packages/@aws-cdk/aws-apigateway/lib/api-definition.ts:92]: sure

**Review by rix0rrr — DISMISSED** [review]

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**